torch/csrc/autograd/engine.cpp

```diff
@@ -288,6 +288,47 @@ static variable_list call_post_hooks(Function& fn, variable_list outputs, variab
   return outputs;
 }
 
+static bool is_compatible_type(const at::Type& expected, const at::Type& actual) {
```

Comment:
Worth a comment saying that it's OK for a dense tensor to have a sparse gradient, but not vice versa.